File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void solve(vector<int>&candidates,vector<int>&output,vector<vector<int>>&ans,int target,int index,int n){
        if(target<0)return;
        if(target==0){
            ans.push_back(output);
            return;
        }
        // include
        for(int i=index;i<n;i++){
            if(i>index && candidates[i]==candidates[i-1])continue;
            output.push_back(candidates[i]);
            solve(candidates,output,ans,target-candidates[i],i+1,n);
            output.pop_back();
        }

    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        int index=0;
        vector<int>output;
        vector<vector<int>>ans;
        int n=candidates.size();
        sort(candidates.begin(),candidates.end());
        solve(candidates,output,ans,target,index,n);
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public void solve(int[] candidates,int target,List<Integer>output,List<List<Integer>>ans,int ind){
          if(target==0){
            ans.add(new ArrayList<>(output));
            return;
          }
          if(target<0)return ;
          for(int i=ind;i<candidates.length;i++){
             if(i>ind && candidates[i]==candidates[i-1])continue;
             if(target-candidates[i]>=0){output.add(candidates[i]);
             solve(candidates,target-candidates[i],output,ans,i+1);
             output.remove(output.size()-1);}
             
          }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<Integer>output=new ArrayList<>();
        List<List<Integer>> ans=new ArrayList<>();
        Arrays.sort(candidates);
        solve(candidates,target,output,ans,0);
        return ans;
    }
}